File: 229. Majority Element II/C++/Solution.cpp
class Solution {
public:
    vector<int> majorityElement(vector<int> &nums) {
        if (nums.size() < 2) {
            return nums;
        }
        int a = nums[0];
        int b = nums[0];
        int countA = 0;
        int countB = 0;
        for (int num : nums) {
            if (num == a) {
                countA++;
            } else if (num == b) {
                countB++;
            } else if (countA == 0) {
                a = num;
                countA = 1;
            } else if (countB == 0) {
                b = num;
                countB = 1;
            } else {
                countA--;
                countB--;
            }
        }
        if (a == b) {
            return vector<int> {a};
        }
        vector<int> result;
        countA = 0;
        countB = 0;
        for (int num : nums) {
            if (a == num) {
                countA++;
            }
            if (b == num) {
                countB++;
            }
        }
        if (countA > nums.size() / 3) {
            result.push_back(a);
        }
        if (countB > nums.size() / 3) {
            result.push_back(b);
        }
        return result;
    }
};

File: 229. Majority Element II/Python/Solution.py
class Solution(object):
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        return self.majority_element(nums, 3)

    def majority_element(self, nums, k):
        counters = {}
        for num in nums:
            if num in counters:
                counters[num] += 1
            elif len(counters.keys()) < k - 1:
                counters[num] = 1
            else:
                for key, count in counters.items():
                    if count == 1:
                        del counters[key]
                    else:
                        counters[key] -= 1
        for num in counters.keys():
            counters[num] = 0
        for num in nums:
            if num in counters:
                counters[num] += 1
        target = len(nums) // k
        result = []
        for num, count in counters.items():
            if count > target:
                result.append(num)
        return result
